presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggIfToAggFilter.java

```diff
@@ -0,0 +1,190 @@
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isAggIfToFilterRewriteEnabled(session);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        ProjectNode sourceProject = captures.get(CHILD);
+
+        Set<AggregationNode.Aggregation> aggregationsToRewrite = aggregationNode.getAggregations().values().stream()
+                .filter(aggregation -> ShouldRewriteAggregation(aggregation, sourceProject))
+                .collect(toSet());
+        if (aggregationsToRewrite.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Extract the VariableReferenceExpressions from these aggregations.
+        // Multiple aggregations may reference the same input. We use a set to dedup them so that we only do the rewrite once per input IF expression.
+        Set<VariableReferenceExpression> aggregationReferences =
+                aggregationsToRewrite.stream().map(a -> (VariableReferenceExpression) a.getArguments().get(0)).collect(toSet());
+
+        // Get the corresponding assignments in the input project.
+        Map<VariableReferenceExpression, RowExpression> sourceAssignments = sourceProject.getAssignments().entrySet().stream()
+                .filter(entry -> aggregationReferences.contains(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        Assignments.Builder newAssignments = Assignments.builder();
+        // We don't remove the IF expression now in case the aggregation has other references to it. These will be cleaned up by other pruning rules.
+        newAssignments.putAll(sourceProject.getAssignments());
+
+        // Map from the aggregation reference to the IF condition reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToConditionReference = new HashMap<>();
+        // Map from the aggregation reference to the IF result reference.
+        Map<VariableReferenceExpression, VariableReferenceExpression> aggregationReferenceToIfResultReference = new HashMap<>();
+
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : sourceAssignments.entrySet()) {
+            VariableReferenceExpression outputVariable = entry.getKey();
+            SpecialFormExpression ifExpr = (SpecialFormExpression) entry.getValue();
+
+            RowExpression condition = ifExpr.getArguments().get(0);
+            VariableReferenceExpression conditionReference = context.getVariableAllocator().newVariable(condition);
+            newAssignments.put(conditionReference, condition);
+            aggregationReferenceToConditionReference.put(outputVariable, conditionReference);
+
+            RowExpression ifResult = ifExpr.getArguments().get(1);
```

Comment:
- s/ifResult/trueResult
- also, we didn't validate if the if expr has 2 or 3 arguments, which we don't support the latter one.

Reply:
shouldRewriteAggregation already validated that the If expression doesn't have else results. So we are good here.